### PR TITLE
feat: email badging on bulk email send

### PR DESCRIPTION
**Pull request by sreehari2003**
## Change Summary

Provide summary of changes with issue number if any.

## Change type

- [x] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [ ] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of

### CODE REVIEW

File: packages/nc-gui/components/dlg/share-and-collaborate/View.vue [hunk 1]
```diff
@@ -95,13 +95,13 @@ watch(showShareModal, (val) => {
     :width="formStatus === 'manageCollaborators' ? '60rem' : '40rem'"
   >
     <div v-if="formStatus === 'project-collaborateSaving'" class="flex flex-row w-full px-5 justify-between items-center py-1">
-      <div class="flex text-base" :style="{ fontWeight: 500 }">Adding Collaborators</div>
+      <div class="flex text-base" :style="{ fontWeight: 500 }">Adding Members</div>
```
Comment (wingkwong) [inline packages/nc-gui/components/dlg/share-and-collaborate/View.vue:98]: don't use `:style` ...

File: packages/nc-gui/components/dlg/share-and-collaborate/View.vue [hunk 2]
```diff
@@ -95,13 +95,13 @@ watch(showShareModal, (val) => {
...
       <a-spin :indicator="indicator" />
     </div>
     <template v-else-if="formStatus === 'project-collaborateSaved'">
       <div class="flex flex-col py-1.5">
         <div class="flex flex-row w-full px-5 justify-between items-center py-0.5">
-          <div class="flex text-base" :style="{ fontWeight: 500 }">Collaborators added</div>
+          <div class="flex text-base" :style="{ fontWeight: 500 }">Members added</div>
```
Comment (wingkwong) [inline packages/nc-gui/components/dlg/share-and-collaborate/View.vue:104]: don't use `:style` ...

File: packages/nc-gui/components/project/AccessSettings.vue [hunk 3]
```diff
@@ -135,6 +135,11 @@ watchDebounced(
   },
 )
 
+const userProjectRole = computed<(typeof ProjectRoles)[keyof typeof ProjectRoles]>(() => {
+  const projectUser = collaborators.value?.find((collab) => collab.id === user.value?.id)
+  return projectUser?.projectRoles
+})
```
Comment (wingkwong) [inline packages/nc-gui/components/project/AccessSettings.vue:141]: what's that for? seems not in use.

File: packages/nc-gui/components/workspace/InviteSection.vue [hunk 4]
```diff
@@ -7,22 +7,80 @@ const inviteData = reactive({
...
+
+const emailValidation = reactive({
+  isError: false,
+  message: '',
+})
+
+const validateEmail = (email: string): boolean => {
+  const regEx = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
+  return regEx.test(email)
+}
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/InviteSection.vue:22]: we already got `validateEmail` in `packages/nc-gui/utils/validation.ts`

File: packages/nc-gui/components/workspace/InviteSection.vue [hunk 5]
```diff
@@ -7,22 +7,80 @@ const inviteData = reactive({
...
 const { workspaceRoles } = useRoles()
 
+// all user input emails are stored here
+const emailBadges = ref<Array<string>>([])
+watch(inviteData, (newVal) => {
+  if (newVal.email.includes(',')) {
+    emailBadges.value.push(newVal.email.split(',')[0])
+    inviteData.email = ''
+  }
+  if (newVal.email.length < 1 && emailValidation.isError) {
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/InviteSection.vue:36]: `newVal.email.length < 1` = `!newVal.email.length`. Only one case -> no need to use range.

File: packages/nc-gui/components/workspace/InviteSection.vue [hunk 6]
```diff
@@ -7,22 +7,80 @@ const inviteData = reactive({
...
+})
+
+const scrollToEnd = () => {
+  if (divRef.value) {
+    divRef.value.scrollLeft = divRef.value.scrollWidth
+  }
+}
+
+const handleEnter = () => {
+  if (inviteData.email.length < 1) {
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/InviteSection.vue:48]: same as above.

File: packages/nc-gui/components/workspace/InviteSection.vue [hunk 7]
```diff
@@ -7,22 +7,80 @@ const inviteData = reactive({
...
 const { isInvitingCollaborators } = storeToRefs(workspaceStore)
 const { workspaceRoles } = useRoles()
 
+// all user input emails are stored here
+const emailBadges = ref<Array<string>>([])
+watch(inviteData, (newVal) => {
+  if (newVal.email.includes(',')) {
+    emailBadges.value.push(newVal.email.split(',')[0])
+    inviteData.email = ''
+  }
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/InviteSection.vue:35]: please explain this logic. why do we only push the first email to emailBadges?

File: packages/nc-gui/components/workspace/InviteSection.vue [hunk 8]
```diff
@@ -7,22 +7,80 @@ const inviteData = reactive({
...
-    await _inviteCollaborator(inviteData.email, inviteData.roles)
+    let inviteEmails = ''
+    emailBadges.value.forEach((el, index) => {
+      // prevent the last email from getting the ","
+      if (index === emailBadges.value.length - 1) {
+        inviteEmails += el
+      } else {
+        inviteEmails += `${el},`
+      }
+    })
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/InviteSection.vue:74]:
```ts
const inviteEmails = emailBadges.value.join(',')
```

File: packages/nc-gui/components/workspace/InviteSection.vue [hunk 9]
```diff
@@ -43,37 +101,68 @@ onMounted(async () => {
...
+          >
+            {{ email }}
+            <component :is="iconMap.close" class="ml-1.5 hover:cursor-pointer" @click="emailBadges.splice(index, 1)" />
+          </span>
+          <input
+            id="email"
+            ref="focusRef"
+            v-model="inviteData.email"
+            :placeholder="emailBadges.length < 1 ? 'Enter emails to send invitation' : ''"
+            class="min-w-30 !focus:border-none !outline-0 !focus:outline-0 !bg-[white] ml-2 mr-3 w-full mt-1"
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/InviteSection.vue:134]: double check if `!` is necessary. `outline-0` should cover focus case already.

**Review by wingkwong — COMMENTED** [review]
Need some steps on how to test it.

**Comment by sreehari2003** [comment]
> Need some steps on how to test it.

we can send email bulk  , when user press space or comma email is converted to a badge like gmail. , u might need to check the EE for  UI view

**Review by dstala — APPROVED** [review]

**Comment by wingkwong** [comment]
Well .. some comments are actually NOT resolved ...

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 10]
```diff
@@ -9,6 +9,78 @@ const inviteData = reactive({
...
+
+const emailValidation = reactive({
+  isError: false,
+  message: '',
+})
+
+const validateEmail = (email: string): boolean => {
+  const regEx = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
+  return regEx.test(email)
+}
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:24]: we already got `validateEmail` in `packages/nc-gui/utils/validation.ts`

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 11]
```diff
@@ -9,6 +9,78 @@ const inviteData = reactive({
...
+    emailBadges.value.push(emailToAdd)
+    inviteData.email = ''
+  }
+  if (newVal.email.length < 1 && emailValidation.isError) {
+    emailValidation.isError = false
+  }
+})
+
+const handleEnter = () => {
+  if (inviteData.email.length < 1) {
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:69]: `!inviteData.email.length`

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 12]
```diff
@@ -9,6 +9,78 @@ const inviteData = reactive({
...
+     **/
+    if (emailBadges.value.includes(emailToAdd)) {
+      insertOrUpdateString(emailToAdd)
+      inviteData.email = ''
+      return
+    }
+    emailBadges.value.push(emailToAdd)
+    inviteData.email = ''
+  }
+  if (newVal.email.length < 1 && emailValidation.isError) {
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:63]: `!newVal.email.length`

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 13]
```diff
@@ -9,6 +9,78 @@ const inviteData = reactive({
...
+  const isNewEmail = newVal.email.charAt(newVal.email.length - 1) === ',' || newVal.email.charAt(newVal.email.length - 1) === ' '
+  if (isNewEmail && newVal.email.trim().length > 1) {
+    const emailToAdd = newVal.email.split(',')[0].trim() || newVal.email.split(' ')[0].trim()
+    if (!validateEmail(emailToAdd)) {
+      emailValidation.isError = true
+      emailValidation.message = 'INVALID EMAIL'
+      return
+    }
+    /** 
+     if email is already enterd we delete the already
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:52]: typo

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 14]
```diff
@@ -31,18 +103,28 @@ const inviteCollaborator = async () => {
...
 
   try {
+    emailBadges.value.forEach((el, index) => {
+      // prevent the last email from getting the ","
+      if (index === emailBadges.value.length - 1) {
+        inviteData.email += el
+      } else {
+        inviteData.email += `${el},`
+      }
+    })
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:113]: this logic can be replaced by 

```ts
inviteData.email += emailBadges.value.join(',')
```

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 15]
```diff
@@ -76,11 +158,60 @@ const copyUrl = async () => {
...
 }
+
+const focusOnDiv = () => {
+  focusRef.value?.focus()
+  isDivFocused.value = true
+}
+
+// remove one email per backspace
+onKeyStroke('Backspace', () => {
+  if (isDivFocused.value && inviteData.email.length < 1) {
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:176]: `!inviteData.email.length`

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 16]
```diff
@@ -76,11 +158,60 @@ const copyUrl = async () => {
...
+const onPaste = (e: ClipboardEvent) => {
+  const pastedText = e.clipboardData?.getData('text')
+  const inputArray = pastedText?.split(',') || pastedText?.split(' ')
+  // if data is pasted to a already existing text in input
+  // we add existingInput + pasted data
+  if (inputArray?.length === 1 && inviteData.email.length > 1) {
+    inputArray[0] = inviteData.email += inputArray[0]
+  }
+  inputArray?.forEach((el) => {
+    if (el.length < 1) {
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:191]: `!el.length`

File: packages/nc-gui/components/project/InviteProjectCollabSection.vue [hunk 17]
```diff
@@ -128,12 +259,44 @@ const copyUrl = async () => {
...
+                  :is="iconMap.close"
+                  class="ml-0.5 hover:cursor-pointer w-3.5 h-3.5"
+                  @click="emailBadges.splice(index, 1)"
+                />
+              </span>
+              <input
+                id="email"
+                ref="focusRef"
+                v-model="inviteData.email"
+                :placeholder="emailBadges.length < 1 ? 'Enter emails to send invitation' : ''"
```
Comment (wingkwong) [inline packages/nc-gui/components/project/InviteProjectCollabSection.vue:290]: `!emailBadges.length`

File: packages/nc-gui/components/workspace/CollaboratorsList.vue [hunk 18]
```diff
@@ -44,18 +44,18 @@ onMounted(async () => {
 
 <template>
   <div class="nc-collaborator-table-container mt-4 mx-6">
-    <WorkspaceInviteSection v-if="workspaceRoles !== WorkspaceUserRoles.VIEWER" />
-    <div class="w-full h-1 border-t-1 border-gray-100 opacity-50 mt-6"></div>
-    <div class="w-full flex flex-row justify-between items-baseline mt-6.5 mb-2 pr-0.25 ml-2">
-      <div class="text-xl">Collaborators</div>
-      <a-input v-model:value="userSearchText" class="!max-w-90 !rounded-md mr-4" placeholder="Search collaborators">
+    <!-- <div class="w-full h-1 border-t-1 border-gray-100 opacity-50 mt-6"></div> -->
```
Comment (wingkwong) [inline packages/nc-gui/components/workspace/CollaboratorsList.vue:47]: remove this line

**Review by wingkwong — COMMENTED** [review]
The error message should be aligned. Some are in all cap for no reason. Also the logic kept repeating. Do it once and use in different place instead.

**Comment by wingkwong** [comment]
@sreehari2003 Please create a follow-up PR for resolving the above-mentioned issues. I don't think this PR is ready to merge at all. Don't just resolve it without any changes - leave a comment in the convo to discuss if you have different pov. Feel free to DM me if you got any problems. Happy to discuss.

**Comment by wingkwong** [comment]
As discussed with Sreehari, the follow-up changes will be done in hub and sync back later.

**Comment by sreehari2003** [comment]
we wont be using the projectCollabPage anymore so I have removed that file

**Pull request merged**